tests/xrt/03_loopback/main.cpp

```diff
@@ -152,99 +148,90 @@ int main(int argc, char** argv)
-            ecmd->data[XLOOPBACK_CONTROL_ADDR_LENGTH_R_DATA/4] = DATA_SIZE; // length
-#else
-            ecmd->data[XLOOPBACK_CONTROL_ADDR_S1_DATA/4] = bo1devAddr; // s1
-            ecmd->data[XLOOPBACK_CONTROL_ADDR_S2_DATA/4] = bo2devAddr; // s2
-            ecmd->data[XLOOPBACK_CONTROL_ADDR_LENGTH_R_DATA/4] = DATA_SIZE; // length
-#endif
-        }
+       // Detecting the first used memory
+       auto topo = xclbin::get_axlf_section(top, MEM_TOPOLOGY);
+       auto topology = reinterpret_cast<mem_topology*>(header.data() + topo->m_sectionOffset);
 
-        std::cout << "Starting kernel..." << std::endl;
+       for (int i=0; i<topology->m_count; ++i) {
+          if (topology->m_mem_data[i].m_used) {
+             first_mem = i;
+             break;
+          }
+       }
 
-        //Send the "start kernel" command.
-        if(xclExecBuf(handle, execHandle)) {
-            std::cout << "Unable to issue xclExecBuf : start_kernel" << std::endl;
-            std::cout << "FAILED TEST\n";
-            std::cout << "Write failed\n";
-            return 1;
-        }
+       if (first_mem < 0)
+           return 1;
 
-        //Wait on the command finish
-        while (xclExecWait(handle,1000) == 0) {
-            std::cout << "reentering wait...\n";
-        };
 
-        //Get the output;
-        if(xclSyncBO(handle, boHandle1, XCL_BO_SYNC_BO_FROM_DEVICE , DATA_SIZE, 0))
-            return 1;
-        char* bo1 = (char*)xclMapBO(handle, boHandle1, false);
+       std::string kname = "loopback";
+       auto kernel = xrtPLKernelOpen(handle, header.data(), kname.c_str());
+
+       auto boHandle2 = xclAllocBO(handle, DATA_SIZE, XCL_BO_DEVICE_RAM, first_mem);
+       char* bo2 = (char*)xclMapBO(handle, boHandle2, true);
+       memset(bo2, 0, DATA_SIZE);
+       std::string testVector =  "hello\nthis is Xilinx OpenCL memory read write test\n:-)\n";
+       std::strcpy(bo2, testVector.c_str());
+
+       if(xclSyncBO(handle, boHandle2, XCL_BO_SYNC_BO_TO_DEVICE , DATA_SIZE,0))
+           return 1;
+
+       auto boHandle1 = xclAllocBO(handle, DATA_SIZE, XCL_BO_DEVICE_RAM, first_mem);
+
 
-        if (std::memcmp(bo2, bo1, DATA_SIZE)) {
+       std::cout << "\nStarting kernel..." << std::endl;
+
+       xrtRunHandle r = xrtKernelRun(kernel, boHandle1, boHandle2,  DATA_SIZE); // kernel, output, input, parameter
+       xrtRunWait(r);
+       xrtRunClose(r);
+
+       //Get the output;
+       if(xclSyncBO(handle, boHandle1, XCL_BO_SYNC_BO_FROM_DEVICE , DATA_SIZE, 0))
+           return 1;
+       char* bo1 = (char*)xclMapBO(handle, boHandle1, false);
+
+       if (std::memcmp(bo2, bo1, DATA_SIZE)) {
             std::cout << "FAILED TEST\n";
             std::cout << "Value read back does not match value written\n";
             return 1;
-        }
+       }
 
         //Clean up stuff
         munmap(bo1, DATA_SIZE);
         munmap(bo2, DATA_SIZE);
```

Comment:
If you replace 'munmap' with 'xclUnmapBO' then this test will almost certainly be working on windows also.

Reply:
Done, Thanks